### PR TITLE
Make MockSingleton multiply instantiable (rebased on #820)

### PR DESCRIPTION
**Pull request by maackle**
I cherry-picked commits from #815 onto `develop` to build this branch, rather than dealing with merge conflicts of a true rebase. It also has one extra commit to fix a mistake.

### CODE REVIEW

**Comment by zippy** [comment]
CI broken...

File: net/src/mock_worker.rs [hunk 1]
```diff
@@ -284,17 +279,24 @@ impl NetWorker for MockWorker {
...
+
                 let (tx, rx) = mpsc::channel();
                 self.mock_msgs.push(rx);
                 mock.register(&app.dna_address, &app.agent_id, tx)?;
                 return Ok(());
             }
         }
-
+        let map_lock = MOCK_MAP.read().unwrap();
+        let mut mock = map_lock.get(&self.network_name).unwrap().lock().unwrap();
```
Comment (ddd-mtl) [inline net/src/mock_worker.rs]: I don't believe the "get" is guaranteed to work so I'd put an expect instead of unwrap here.

**Comment by maackle** [comment]
So far I'm stumped why the last app_spec test fails here. It passed on develop. This is a one-file change. Does anyone have an idea?

**Comment by maackle** [comment]
I wrote #835 which changes an aspect of container network config initialization and makes this PR work. I rebased on that PR and force-pushed.

**Comment by maackle** [comment]
@neonphog yeah I'm punting on shutdown. Shutdown would get us to the point of running independent tests in nodejs, but we still need multiple mock networks if we want parallel tests, and multiple mocks also gets us to independent tests (I hope)

**Comment by maackle** [comment]
to david's last point, we picked up the conversation here: https://github.com/holochain/holochain-rust/issues/841

**Pull request merged**